Computation of dependency graph for strings [TG-2605]

romainbrenguier opened the pull request:
This add a class to compute dependencies between strings:
for instance if we have `s1 = s2.concat(s3)` then s1 depends on s2 and s3.
This does not change the behaviour of the solver: the functionality is only used in a unit test and under `#ifdef DEBUG` in the solver for giving the graph as debug information.
Ultimately we want to use this dependency graph to minimise the number of `string_constraintst` that are created and instantiated in the underlying solver.
For now only `concat` and `insert` are implemented, and for all other functions we over-approximate the dependencies (for instance for `s1 = s2.substring(10)` the graph right now has both s1 depends on s2 and s2 depends on s1, though actually only s1 depends on s2).

romainbrenguier committed:
Correct string-max-input-length tests

Remove test from max_input_length

This remove test which was using the way max-input-length was limiting
the size of strings, even constant one.
The solver is now more permisives and the long strings would just get
troncated instead of leading to a solver contradiction.

For MemberTest, we add a new test.
We now have two tests with different values of the option, which should
give different results.

romainbrenguier committed:
Create string_refinement_util for utility function

string_refinement.cpp file is becoming too large.
This groups utility classes used by string_refinement in new files to
reduce string_refinement.cpp a bit.

romainbrenguier committed:
Make interface of equation_symbol_mappingt clearer

Make the fields private, document methods and put definition in cpp file.

romainbrenguier committed:
Unit tests for dependency graph

romainbrenguier committed:
Class for string builtin functions and dependences

String builtin functions can be initialized from
function_application_exprt for some function names, for now only
cprover_string_concat_func and cprover_string_insert_func are supported
but this could be extended to all builtin functoins supported by the
solver.

We use information of string argument and return values of the
functions for creating a dependency graph between the strings, stored
in the string dependences structure.

We also define indexes in the dependency graph so that we can reuse
generic algorithms of util/graph.

romainbrenguier committed:
Pull symbol generation out of constraint generator

romainbrenguier committed:
Pull array_pool class out of constraint generator

Pull out this part of constraint generator to decrease the size of the
class, make unit test easier and the code more modular and reusable.

romainbrenguier committed:
Move utility functions to string_refinement_util

This reduces the size of string_refinement files and seperate functions
that could be reused outside of string_refinement, for instace for unit testing.

romainbrenguier committed:
Pull out a get_function_name function

This decrease the size of add_axioms_for_function_application and this
new function can be reused.

romainbrenguier committed:
Define function for array_pointer_association

This allows the constraint generator to do a pass on equations without
adding axioms but just adding the assaciation between pointer and arrays.

romainbrenguier committed:
Separate pointer/function substitutions in solver

We make the solver do two passes on the equations:
there is now one pass just for pointer and array associations before
substituting function applications.

tautschnig commented on src/solvers/refinement/string_constraint_generator.h:
```diff
@@ -27,6 +27,53 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+private:
+  // associate arrays to char pointers
+  std::map<exprt, array_string_exprt> arrays_of_pointers;
```
Is there a reason not to use a `std::unordered_map` here? And indeed in all places except for those that require lexicographic sorting or stable iterators?

romainbrenguier replied:
No particular reason, I can change that

tautschnig commented on src/solvers/refinement/string_constraint_generator_insert.cpp:
```diff
@@ -74,8 +74,9 @@ exprt string_constraint_generatort::add_axioms_for_insert(
+  const array_string_exprt res =
+    array_pool.find(f.arguments()[1], f.arguments()[0]);
+  // \todo: is this really 3 and not 4 ?
   const exprt &offset = f.arguments()[3];
```
Isn't there already an array at index 4? Having both offset and array at that place would seem surprising?

romainbrenguier replied:
I was thinking the two were swapped because in the documentation of the function the order is different, but actually it's the documentation that needs to be updated.

tautschnig commented on src/solvers/refinement/string_refinement.h:
```diff
@@ -27,57 +27,11 @@ Author: Alberto Griggio, alberto.griggio@gmail.com
 #include <solvers/refinement/string_constraint.h>
 #include <solvers/refinement/string_constraint_generator.h>
 #include <solvers/refinement/string_refinement_invariant.h>
+#include <solvers/refinement/string_refinement_util.h>
```
Do you actually need this in the header file?

romainbrenguier replied:
Not easy to avoid because some fields have type defined there.

tautschnig replied:
If it's unavoidable then that's what it is, but if forward declarations would do the trick (as may be the case with the other includes) then it would help reducing dependencies during compilation. Just something to keep in mind.

tautschnig commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -0,0 +1,315 @@
+/// Keep track of dependencies between strings.
+/// Each string points to primitive calls on which it depends,
+/// each primitive points to the strings on which the result depend.
+class dependency_grapht
```
Is there a really good reason not to use/inherit from `grapht`?

romainbrenguier replied:
We have two different types of nodes (`primitive_nodet` and `string_nodet`), and we do not need to store information on edges, so using `grapht` seems like more work.

smowton replied:
I think you should reconsider; grapht will provide you with useful graph algorithms "off the peg" which you will surely want in time. Either encode your two node types as a class hierarchy (so you can use dynamic_cast to determine the type of a given node) or as a more- or less-formal union type.

tautschnig replied:
Sounds like a slightly weird sort of graph (different nodes, no edges) :-)  I just want to make sure we keep the amount of duplicate code to a minimum. Not-invented-here syndrome is a pretty dangerous disease.

romainbrenguier replied:
@smowton 
> I think you should reconsider; grapht will provide you with useful graph algorithms "off the peg" which you will surely want in time. Either encode your two node types as a class hierarchy (so you can use dynamic_cast to determine the type of a given node) or as a more- or less-formal union type.

Not sure I will need these algorithms, since this is an acyclic graph, computing SCCs is useless, so is `is_dag`, and `shortest_loop`. `shortest_path` is not really meaningful in this context. Maybe `get_reachable` would make sense but anyway I don't like to have to inherit from a class to be able to use its algorithms. I'd prefer the algorithms to be implemented as function templates, here for instance for `get_reachable`: https://github.com/romainbrenguier/cbmc/pull/5

tautschnig replied:
Please do bring in changes like https://github.com/romainbrenguier/cbmc/pull/5 - otherwise "you're not doing it my way so I don't use it" sounds like not-invented-here syndrome has already manifested. The code base is pretty big already, and someone will have to continue to maintain all these copies of almost-the-same-but-not-quite-the-same code.

romainbrenguier replied:
@tautschnig I'm working on it. I guess only the `output_dot` is necessary for this pull request since it's the only function that is duplicated. Or should I generalise the functions that I define to make them usable by grapht?

tautschnig replied:
I might be misunderstanding, so please correct me if I'm wrong: it seems you suggest to retain your own data structure, but intend to use functions that otherwise are only used on `grapht` (moving them from member functions to free functions)?

If the above is right, the best possible picture is making all functions that work on those graph-like objects apply to anything that shares the same API as `grapht`.

romainbrenguier replied:
@tautschnig 
> I might be misunderstanding, so please correct me if I'm wrong: it seems you suggest to retain your own data structure, but intend to use functions that otherwise are only used on grapht (moving them from member functions to free functions)?

That's right

> If the above is right, the best possible picture is making all functions that work on those graph-like objects apply to anything that shares the same API as grapht.

The public interface of `grapht` is huge so I don't wan't to re-implement all these, just so that maybe one day I can reuse one basic function like `get_reachable`.

chrisr-diffblue replied:
If the public interface of `grapht` is so large as to hinder its reuse, is there any milage in splitting up it's interface into a finer grained class hierarchy?

LAJW replied:
All computer applications consist of graphs. But each has different needs; some are trees, some are lists, some are directed (like this one), some are not and have edges (like `grapht`). Directed graphs can be easily expressed with `shared_ptr`s.

`grapht` has too many unnecessary features for it to be useful here, and we shouldn't modify it just to suit this one particular use case.

peterschrammel replied:
Now, we have a `dependence_grapht` and a `dependency_grapht` in the code base. I don't think that this a good idea.

peterschrammel replied:
This looks quite specialised; still, it could be based on `grapht` in the way @smowton suggests. 
In particular, I don't understand that argument:

> The public interface of grapht is huge so I don't wan't to re-implement all these

Why would you need to reimplement any of these?

peterschrammel replied:
How is the graph going to be used? At the moment, it seems you can only print it.
Maybe this will shed some light on which algorithms from `grapht` you'll reuse/reimplement.

smowton replied:
@chrisr-diffblue you would never reimplement its public interface; the typical usage looks like this: https://github.com/diffblue/cbmc/blob/develop/src/analyses/call_graph.h#L113

Basically you define an edge struct (or don't if, like this graph, edges are unlabelled), then a node struct (here a tagged union of intrinsic function calls and strings), then finally inherit grapht<my_node_typet>. It gives you standard mutators and accessors for making nodes, adding / querying / deleting edges, and so on, plus a small library of graph algorithms that work over a grapht.

For this use case you'd do something like
```
struct string_dependence_graph_nodet : graph_nodet {
  exprt string_or_intrinsic; // Use .id() to differentiate them if context isn't enough
};
struct string_dependence_grapht : grapht<string_dependence_graph_nodet>
{
} mygraph;

auto new_node_idx = mygraph.add_node();
mygraph[new_node_idx].string_or_intrinsic = some_expr;
...
```

The advantage is that that you get things like `mygraph.get_reachable(some_node_idx, forwards)` for free, and so avoid writing yet another common graph algorithm implementation for each type that needs it.

I agree that this sort of thing should be done as free functions expecting a concept, a la STL algorithms, but it doesn't make a massive difference to how you use them -- you'd still need to conform to an interface somehow by implementing `get_node`, `get_edges` and similar, probably by inheriting some useful mixin.

Whether this use case will ever **want** the standard graph algorithms though? @romainbrenguier thinks no, I think we might as well describe this as a generic graph just in case we want them in the future since it's little effort to do so, but I don't have a strong enough opinion on the matter to block on this.

romainbrenguier replied:
@peterschrammel 
> In particular, I don't understand that argument:
> > The public interface of grapht is huge so I don't wan't to re-implement all these

> Why would you need to reimplement any of these?

This wasn't an argument against inheriting from grapht, but against "making all functions that work on those graph-like objects apply to anything that shares the same API as grapht".

romainbrenguier replied:
@peterschrammel 
> Now, we have a dependence_grapht and a dependency_grapht in the code base. I don't think that this a good idea.

What about `string_dependence_grapht` or just `string_dependencest`?

peterschrammel replied:
@romainbrenguier, either name is fine.

How is the graph going to be used?

romainbrenguier replied:
@peterschrammel 
We will look at the set of primitives that are "relevant" for the solver: that is all test and comparison primitives and their dependencies. These will get added as string_constraints but not the others (the remaining primitives do not affect satisfiability). The algorithm we need for that would be some kind of `get_reachable` which takes a set of nodes as inputs instead of a single node. I can write that as a function template in `util/graph` so that it can be reused by any graph-like structure.

martin-cs commented:
Sounds like a good idea in general; particularly if you can push ITEs through various string constraints.  @romainbrenguier what kind of feedback do you want and when?

romainbrenguier commented:
@martin-cs Ideally I'd like feedback before the end of the week so that I can start working on the next step on Monday. The idea will be to actually use the graph and drop constraints that do not affect satisfiability.
I'm open to any feedback, but there may be some changes soon when I start using the graph, so don't worry about the small details, more about the global structure.

smowton commented on src/solvers/refinement/string_constraint_generator_main.cpp:
```diff
@@ -247,12 +245,32 @@ string_constraint_generatort::associate_char_array_to_char_pointer(
+  const exprt &length = array_expr.length();
+  if(length == infinity_exprt(length.type()))
+  {
+    auto pair = length_of_array.insert(
```
Is it possible that the same array is inserted more than once here? If so should we check that the length we're about to set matches the one already stored?

romainbrenguier replied:
If the same array is added twice then the invariant at the end of this function should fail.

smowton commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -0,0 +1,315 @@
+    bool depends_on_unknown_primitive = false;
+  };
+
+  string_nodet &get_node(const array_string_exprt &e);
```
`get_node` and `make_node` seem like bad names to distinguish the two node types

romainbrenguier replied:
They are different in behaviour in that `make_node` always create a new node, whereas `get_node` first looks at whether it exists or not, otherwise I would have given both function the same name.

smowton commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -0,0 +1,315 @@
+};
+
+/// Keep track of dependencies between strings.
+/// Each string points to primitive calls on which it depends,
```
What is a primitive call? You mean one of the intrinsics provided by the string solver, like __CPROVER_string_equal et al?

romainbrenguier replied:
yes, but I'm open to suggestion for a better vocabulary.

smowton replied:
I'd probably call them "string intrinsics", since "primitives" makes me think of primitive types

romainbrenguier replied:
@smowton What about "builtin functions" ? I would rename `string_primitivet` to `string_builtin_functiont` and so on.

smowton replied:
They're not quite builtins, they belong to the string solver. So `string_solver_function_callt`?

romainbrenguier committed:
Define function template for output_dot

This makes output_dot more reusable as we don't have to inherit from
grapht to use it.

peterschrammel commented on unit/solvers/refinement/string_refinement/dependency_graph.cpp:
```diff
@@ -0,0 +1,195 @@
+        });
+      }
+#endif
+      /*
```
use `//` comments

peterschrammel commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -9,7 +9,11 @@
 #define CPROVER_SOLVERS_REFINEMENT_STRING_REFINEMENT_UTIL_H
 
+#include <algorithm>
+#include <iostream>
```
only include that in the .cpp file

peterschrammel commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -150,4 +154,162 @@ class equation_symbol_mappingt
+  void
+  add_dependency(const array_string_exprt &e, const primitive_nodet &primitive);
+
+  /// Mark node for `e` as dependending on unknown primitive
```
depending

peterschrammel commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -150,4 +154,162 @@ class equation_symbol_mappingt
+/// \return true if a node was added, if false is returned it either means that
+///   the right hand side is not a function application
+/// \todo there should be a class with just the three functions we require here
+bool add_node(
```
please describe the params

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -283,107 +283,6 @@ static void substitute_function_applications_in_equations(
     eq.rhs() = substitute_function_applications(eq.rhs(), generator);
 }
 
-/// For now, any unsigned bitvector type of width smaller or equal to 16 is
```
Please fix the typo in the commit message `string_refiniment_util`

romainbrenguier replied:
fixed

allredj commented on src/solvers/refinement/string_constraint_generator.h:
```diff
@@ -27,6 +27,60 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
+  array_string_exprt find(const refined_string_exprt &str);
+
+  /// Argument are passed to primitive as a struct with a pointer and length
+  /// field, this convert this struct to an array.
```
`convert` -> `converts`

allredj commented on src/solvers/refinement/string_constraint_generator_main.cpp:
```diff
@@ -247,12 +244,32 @@ string_constraint_generatort::associate_char_array_to_char_pointer(
+}
+
 /// Associate a char array to a char pointer.
 /// Insert in `arrays_of_pointers_` a binding from `ptr` to `arr`.
```
`arrays_of_pointers_` has been renamed

allredj replied:
And also `length_of_array_` below.

allredj commented on src/solvers/refinement/string_constraint_generator_main.cpp:
```diff
@@ -273,21 +290,8 @@ exprt string_constraint_generatort::associate_array_to_pointer(
-    it_bool.second, "should not associate two arrays to the same pointer");
-  add_default_axioms(to_array_string_expr(array_expr));
+  array_pool.insert(pointer_expr, array_expr);
+  add_default_axioms(array_expr);
```
I think the documentation of the method needs to be updated according to the new behaviour.

allredj commented on src/solvers/refinement/string_constraint_generator_main.cpp:
```diff
@@ -401,18 +405,27 @@ exprt string_constraint_generatort::add_axioms_for_constrain_characters(
+
+array_string_exprt array_poolt::of_argument(const exprt &arg)
+{
+  const auto structure = expr_checked_cast<struct_exprt>(arg);
```
Is `structure` the most representative name we can find for that?

allredj commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -75,6 +75,22 @@ bool has_char_array_subexpr(const exprt &expr, const namespacet &ns)
   return false;
 }
 
+/// Argument of string primitives are passed as struct containing a
```
Doc string is weird...

allredj commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -75,6 +75,22 @@ bool has_char_array_subexpr(const exprt &expr, const namespacet &ns)
 }
 
+/// Argument of string primitives are passed as struct containing a
+optionalt<refined_string_exprt> get_char_array_of_argument(const exprt &expr)
```
I don't really understand why `argument ` is here. What kind of arguments are we talking about? Can we maybe find a more general name? The parameter is `exprt` so it's not clear what an "argument" actually is.

allredj commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -283,6 +283,23 @@ static void substitute_function_applications_in_equations(
     eq.rhs() = substitute_function_applications(eq.rhs(), generator);
 }
 
+/// Fill the array_pointer correspondance and replace the right hand sides of
```
`correspondance` -> `correspondence`

allredj commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -297,6 +297,45 @@ class dependency_grapht
+  /// Even indexes correspond to primitive nodes, odd indexes to string nodes.
+  typedef std::size_t node_indext;
+
+  node_indext size() const
```
What does that size represent and why is it the sum of both counts?

allredj commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -297,6 +297,45 @@ class dependency_grapht
+    return 2 * p.index;
+  }
+
+  node_indext node_index(const array_string_exprt &s) const
```
What is `node_index`? and why is it a function of the size of `primitive_nodes`?

allredj reviewed (commented):
Looks good! A few questions for my understanding.

romainbrenguier commented:
@peterschrammel @allredj I think I addressed all your comments. Can you re-review?

peterschrammel commented:
There's a compilation error: https://travis-ci.org/diffblue/cbmc/jobs/350242446#L1219

peterschrammel commented on src/solvers/refinement/string_constraint_generator_main.cpp:
```diff
@@ -87,8 +86,8 @@ constant_exprt string_constraint_generatort::constant_char(
-symbol_exprt string_constraint_generatort::fresh_symbol(
-  const irep_idt &prefix, const typet &type)
+symbol_exprt symbol_generatort::
+operator()(const irep_idt &prefix, const typet &type)
```
(that's a weird line break)

peterschrammel commented on src/solvers/refinement/string_constraint_generator_main.cpp:
```diff
@@ -247,16 +244,35 @@ string_constraint_generatort::associate_char_array_to_char_pointer(
-    arrays_of_pointers_.insert(std::make_pair(char_pointer, array_sym));
+    arrays_of_pointers.insert(std::make_pair(char_pointer, array_sym));
   array_string_exprt result = to_array_string_expr(insert_result.first->second);
+#if 0
```
Please add a comment why this is deactivated

romainbrenguier replied:
actually I'm going to remove this line, it cannot be activated, and is unnecessary (default axioms are added on fresh_strings, for the already existing strings, constraints are unnecessary)

peterschrammel commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -0,0 +1,414 @@
+bool is_char_type(const typet &type)
+{
+  return type.id() == ID_unsignedbv &&
+         to_unsignedbv_type(type).get_width() <= 16;
```
Is this constant language-specific?

romainbrenguier replied:
The constant is documented in the header. For now we consider the solver only works for strings whose characters are at most 16 bits so that's fine for java and C. Ideally it should be a parameter of the solver or activated for all primitive types.

tautschnig replied:
Use `util/magic.h`. Documentation cannot be read by machines and doing `grep` for arbitrary terms is hard.

peterschrammel commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -0,0 +1,414 @@
+
+bool has_char_array_subexpr(const exprt &expr, const namespacet &ns)
+{
+  for(auto it = expr.depth_begin(); it != expr.depth_end(); ++it)
```
I'd personally prefer braces around the multi-line `for` block.

romainbrenguier replied:
I'm replacing it by a `find_if` which is more appropriate

peterschrammel reviewed (commented):
It's still not clear to me how you are going to query the dependency graph in the solver and why the functions provided by `grapht` are not useful for performing these queries.

romainbrenguier commented:
@peterschrammel 
> It's still not clear to me how you are going to query the dependency graph in the solver and why the functions provided by grapht are not useful for performing these queries.

A beginning of answer on how we are going to use it: https://github.com/romainbrenguier/cbmc/pull/6
I'm not saying the grapht functions are not useful, just that we don't need to inherit from grapht to use them.

peterschrammel reviewed (approved):
Let's go ahead with this. I hope the whole picture becomes clearer when we see `string_dependencest` in action in the next PR.

romainbrenguier committed:
Remove default axiom in associate array to pointer

These default axioms are not useful, as constraints are already added on
input strings and fresh strings.
This also moves default axoims from associate_char_array_of_pointer to
char_array_of_pointer, because the first function will be pulled out of
the class.
